packages/charts/react-charting/src/components/GroupedVerticalBarChart/GroupedVerticalBarChart.base.tsx

```diff
@@ -437,82 +426,82 @@ export class GroupedVerticalBarChartBase
     const singleGroup: JSX.Element[] = [];
     const barLabelsForGroup: JSX.Element[] = [];
 
-    const tempDataSet = Object.keys(this._datasetForBars[0]).splice(0, this._keys.length);
-    tempDataSet.forEach((datasetKey: string, index: number) => {
-      const refIndexNumber = singleSet.indexNum * tempDataSet.length + index;
-      const pointData = singleSet[datasetKey];
-      if (pointData) {
-        const yBarScale = pointData.useSecondaryYScale && yScaleSecondary ? yScaleSecondary : yScalePrimary;
+    this._legends.forEach((legendTitle: string, legendIndex: number) => {
+      const barPoints = singleSet[legendTitle];
+      if (barPoints) {
+        const yBarScale = barPoints[0].useSecondaryYScale && yScaleSecondary ? yScaleSecondary : yScalePrimary;
 
         // To align the centers of the generated bandwidth and the calculated one when they differ,
         // use the following addend.
-        const xPoint = xScale1(datasetKey) + (xScale1.bandwidth() - this._barWidth) / 2;
-        const yPoint = Math.max(yBarScale(pointData.data), 0);
-        let startColor = pointData.color ? pointData.color : getNextColor(index, 0, this.props.theme?.isInverted);
+        const xPoint = xScale1(legendTitle) + (xScale1.bandwidth() - this._barWidth) / 2;
+        const isLegendActive = this._legendHighlighted(legendTitle) || this._noLegendHighlighted();
+        const barOpacity = isLegendActive ? '' : '0.1';
+        const gradientId = getId('GVBC_Gradient') + `_${singleSet.indexNum}_${legendIndex}`;
+        let startColor = barPoints[0].color;
         let endColor = startColor;
 
         if (this.props.enableGradient) {
-          startColor = pointData.gradient?.[0] || getNextGradient(index, 0, this.props.theme?.isInverted)[0];
-          endColor = pointData.gradient?.[1] || getNextGradient(index, 0, this.props.theme?.isInverted)[1];
-          pointData.color = startColor;
+          startColor = barPoints[0].gradient![0];
+          endColor = barPoints[0].gradient![1];
```

Comment:
>gradient![1] [](http://example.com/codeflow?start=34&length=12)

this value can be null as well. In that case we will have to use getNextGradient.

Reply:
I see this has been already handled later